### PR TITLE
Manual - Extending Oakestra

### PR DESCRIPTION
**Pull request by melkodary**
- [x] Addons
- [x] Hooks
- [x] Custom resources

### CODE REVIEW

File: content/docs/manuals/extending-oakestra/create_addon.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+```
+
+Build and push the Docker image to a container registry, such as Docker Hub:
+```bash
+docker build -t your-username/addon-name:latest .
+docker push your-username/addon-name:latest
+```
+
+### Publish the Addon to the Marketplace
+Define the addon configuration.</br>
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/create_addon.md]: remove `</br>`

File: content/docs/manuals/extending-oakestra/create_addon.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+draft: false
+weight: 1
+toc: true
+seo:
+  title: "" # custom title (optional)
+  description: "" # custom description (recommended)
+  canonical: "" # custom canonical URL (optional)
+  noindex: false # false (default) or true
+---
+
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/create_addon.md:13]: There is a short description needed on "Addons" and what can you package as an addon. Highlight why this is a powerful functionality that makes Oakestra future-proof

File: content/docs/manuals/extending-oakestra/install_addon.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+---
+title: "Install an Addon"
+summary: ""
+draft: false
+weight: 2
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/install_addon.md]: Maybe "install addons" should come before "create addons" as most people would only want to use existing addons? Also this documentation is written with basic Addon description that should be read first. 

If yes, please swap the weight of create and install docs

File: content/docs/manuals/extending-oakestra/install_addon.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+Addons in Oakestra allow developers to extend the platform by adding new features or enhancing existing ones. This guide focuses on the installation process for addons, enabling you to integrate published addons into your Oakestra environment seamlessly.
+
+
+## Install an Addon
+
+{{< callout context="note" title="Note" icon="outline/info-circle" >}}
+Before installing an Addon make sure that the Addons System is running and that the Addon to be installed is published in the Marketplace.
+
+- Verify that the Addon is published in the Marketplace. This can be done by sending a `GET` request to `/api/v1/marketplace/addons/{addon_marketplace_id}`.
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/install_addon.md]: Is there a public addon marketplace that will continue to run. How will someone get to know the "marketplace_id"?

File: content/docs/manuals/extending-oakestra/install_addon.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+- Also verify that the Addons Engine component is running inside your orchestrator.
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/install_addon.md]: How to verify this? Give the command or point to specific pointer in setup that someone should lookout for

File: content/docs/manuals/extending-oakestra/install_addon.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+{{< /callout >}}
+
+Installing an addon is a simple step involving sending a `POST` request to the Addons Engine - `/api/v1/addons`. Example of the body of the request could be:
+```json
+{
+  "marketplace_id": "{addon_marketplace_id}"
+}
+```
+
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/install_addon.md:28]: make a link-card and point to "installing apps using API" documentation with the title: "Want to learn how to use Oakestra API? See our detailed documentation."

File: content/docs/manuals/extending-oakestra/hooks.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+weight: 4
+toc: true
+seo:
+  title: "" # custom title (optional)
+  description: "" # custom description (recommended)
+  canonical: "" # custom canonical URL (optional)
+  noindex: false # false (default) or true
+---
+
+## Introduction
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/hooks.md]: No need for the title.

File: content/docs/manuals/extending-oakestra/hooks.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+Hooks in Oakestra provide a mechanism for listening to and reacting to system life-cycle events. By leveraging hooks, developers can decouple components and enable services to dynamically respond to events like creation, updates, or deletion of system entities. This section explains how to use hooks effectively, with step-by-step instructions for registering and implementing them.
+
+Hooks allow developers to:
+1. **Subscribe to Events**: Register services to listen for system life-cycle events such as entity creation, updates, or deletion.
+2. **React to Events**: Perform specific actions when events occur, either synchronously (blocking) or asynchronously (non-blocking).
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/hooks.md]: Can you package this within a "tip callout" (https://getdoks.org/docs/basics/shortcodes/#custom-callout-titles)? You can use the title "Why should developers use hooks?"

File: content/docs/manuals/extending-oakestra/custom_resources.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+weight: 5
+toc: true
+seo:
+  title: "" # custom title (optional)
+  description: "" # custom description (recommended)
+  canonical: "" # custom canonical URL (optional)
+  noindex: false # false (default) or true
+---
+
+## Introduction to Custom Resources
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/custom_resources.md]: No need for the title

File: content/docs/manuals/extending-oakestra/custom_resources.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+Custom Resources in Oakestra provide a mechanism to extend the system’s API dynamically by introducing new resource types. Inspired by Kubernetes’ Custom Resources, this feature allows developers to define and manage tailored data structures to meet specific needs, enabling Oakestra to adapt to a wide range of scenarios.
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/custom_resources.md]: Link to Kubernetes CR documentation

File: content/docs/manuals/extending-oakestra/custom_resources.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+This section demonstrates how to create, manage, and utilize Custom Resources within Oakestra.
+
+Custom Resources in Oakestra enable:
+1. **Dynamic API Extensions**: Add new resource types to the system beyond the default ones.
+2. **Declarative Management**: Store and retrieve structured data through API endpoints.
+3. **Custom Logic**: Use controllers to act on the data defined by the Custom Resources, enabling automation and state management.
+
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/custom_resources.md]: Also a "tip callout" with title "why should you use custom resources in Oakestra?"

File: content/docs/manuals/extending-oakestra/custom_resources.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+#### Hook Integration
+Custom Resources can leverage hooks to trigger synchronous or asynchronous notifications for life-cycle events:
+- **Example Hook**: Notify an external service whenever a resource is created or updated.
+
+---
+
+## Example: Custom Resource Workflow
+
+### Defining a Custom Resource
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/custom_resources.md:91]: Replace with more example oriented titles

File: content/docs/manuals/extending-oakestra/custom_resources.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+- Remove resources when the Custom Resource is deleted.
+
+#### Hook Integration
+Custom Resources can leverage hooks to trigger synchronous or asynchronous notifications for life-cycle events:
+- **Example Hook**: Notify an external service whenever a resource is created or updated.
+
+---
+
+## Example: Custom Resource Workflow
+
```
Comment (nitindermohan) [inline content/docs/manuals/extending-oakestra/custom_resources.md:90]: Can you add a short description of what the example will show?

**Review by nitindermohan — CHANGES_REQUESTED** [review]
@melkodary the PR looks good but there are still some clarity and cosmetic changes required.

File: content/docs/manuals/extending-oakestra/create_addon.md (hunk 2 above)
Comment (melkodary) [reply content/docs/manuals/extending-oakestra/create_addon.md:13]: The suggestion feels more like a `concept` thing than a manual, no?

File: content/docs/manuals/extending-oakestra/prerequisite.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+- Addons Engine
+- Addons Marketplace
+
+<!-- Fix links in another issue - once concepts is merged -->
+You can find more details [here](../../concepts/addons) 
+
+Navigate to director `run-a-cluster` in home folder of the project, run this command
+```bash
+docker compose -f 1-DOC.yaml -f override-addons.yaml up addons_manager addons_monitor marketplace_manager
+```
```
Comment (giobart) [inline content/docs/manuals/extending-oakestra/prerequisite.md]: This command does not match the new startup procedure, which does not require the repository clone. May I suggest the following change?

```suggestion

To use the addons engine and the marketplace, you need to enable them before starting up your Oakestra root orchestrator. To enable these components at cluster startup, you need to export the following *override file* before using the startup command:
>```bash
> export OVERRIDE_FILES=override-addons.yaml
>```

{{< callout context="note" title="Example" icon="outline/info-circle"  >}}

To start a single machine root and cluster orchestrator with the addons engine and the marketplace enabled, you can use:

 >```bash
 > export OVERRIDE_FILES=override-addons.yaml
 > curl -sfL oakestra.io/getstarted.sh | sh - 
 > ```

For a standalone root orchestrator with the addons engine and marketplace enabled, you can use:

> ```bash
> export OVERRIDE_FILES=override-addons.yaml
> curl -sfL https://raw.githubusercontent.com/oakestra/oakestra/develop/scripts/StartOakestraRoot.sh | sh - 
>```

{{< /callout >}}

{{< callout context="tip" title="Did you know?" icon="outline/rocket" >}}
The environment variable `OVERRIDE_FILES` is used at Oakestra startup to customize the services composing the control plane. These files are simple Docker Compose files integrated into Oakestra via the  [override](https://docs.docker.com/compose/how-tos/multiple-compose-files/merge/) functionality.
{{< /callout >}}

```

File: content/docs/manuals/extending-oakestra/prerequisite.md [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+This should spin up 3 docker containers inside the `root-orchestrator`. Notice that also it will run the addons marketplace locally. Currently there isn't a global `Marketplace` for addons. However, in the future there should be one, in which case there wouldn't be a need to run it locally.
+
+{{< callout context="tip" title="Can Addons Engine Run inside a cluster orchestrator too" icon="outline/rocket" >}}
```
Comment (giobart) [inline content/docs/manuals/extending-oakestra/prerequisite.md]:
```suggestion
{{< callout context="tip" title="Can Addons Engine Run inside a cluster orchestrator?" icon="outline/rocket" >}}
```

File: content/docs/manuals/extending-oakestra/install_addon.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+  title: "" # custom title (optional)
+  description: "" # custom description (recommended)
+  canonical: "" # custom canonical URL (optional)
+  noindex: false # false (default) or true
+---
+
+{{< callout context="caution" title="Requirements" icon="outline/alert-triangle" >}}
+Before installing an Addon make sure that the Addons System is running and that the Addon to be installed is published in the Marketplace.
+
+- Verfy that the Addons Engine component is running inside your orchestrator. Visit the [prerequisite section](./prerequisite.md) for more details onto running the Addons Engine component.
```
Comment (giobart) [inline content/docs/manuals/extending-oakestra/install_addon.md]:
```suggestion
- Verfy that the Addons Engine component is running inside your orchestrator. Visit the [prerequisite section](../setup-addons-system) for more details onto running the Addons Engine component.
```

**Review by giobart — CHANGES_REQUESTED** [review]

File: content/docs/manuals/extending-oakestra/prerequisite.md (hunk 14 above)
Comment (melkodary) [reply content/docs/manuals/extending-oakestra/prerequisite.md]: Makes sense! Thanks Giovanni - will resolve this tonight.

**Comment by Mjaethers** [comment]
This branch was rebased onto concepts because otherwise it would be tricky to verify that all the links work

**Comment by nitindermohan** [comment]
@Mjaethers @melkodary is this PR ready for review? I still see that the fig is not compatible with the light/dark mode

**Comment by Mjaethers** [comment]
From my end it's ready for review now

File: content/docs/concepts/oakestra-extensions/addons.md [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+Addons in Oakestra are modular components that extend the platforms core functionality while keeping the foundational architecture lightweight. This design ensures that Oakestra remains efficient, avoiding unnecessary bloat in edge computing environments.
+
+Addons system is composed of two subsystems:
+- **Addons Marketplace**: Consists of Marketplace Manager and MongoDB client.
+- **Addons Engine**: Consists of Addons Manager, Addons Monitor, and MongoDB Client.
+
+## Addons Marketplace
+
+{{<svg "marketplace">}}
```
Comment (nitindermohan) [inline content/docs/concepts/oakestra-extensions/addons.md]: scale is too big. try `svg-smaller`

```suggestion
{{<svg-smaller "marketplace">}}
```

File: content/docs/concepts/oakestra-extensions/addons.md [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+The addons marketplace allows developers to publish their addons for Oakestra as docker images. The marketplace subsystem offers API
+endpoints for developers to publish their addons. Upon receiving a request for registering an addon, the marketplace asynchronously 
+checks if the image is valid. Once it’s verified, the addon’s state in the database changes from `under_review` to `approved` or `failed_verification`.
+
+## Addons Engine
+
+{{<svg "engine">}}
```
Comment (nitindermohan) [inline content/docs/concepts/oakestra-extensions/addons.md]:
```suggestion
{{<svg-smaller "engine">}}
```

File: content/docs/concepts/oakestra-extensions/addons.md [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+## Addon Lifecycle Management
+
+The lifecycle of an addon, from creation to deployment, is managed by the **addons engine** and the **addons marketplace**. This design allows for **hot-swapping**, where new features can be added or swapped without disrupting core services.
+
+When installing an addon, the Addons Engine checks if a similar core component is running. If that's the case, the core component is stopped and the addon is run in its place. In this scenario the addon inherits the port mappings and network specifications of the core component. If no core component exists then it's an extension and no component is replaced.
+
+{{< link-card
+  title="Installing addons"
+  description="Read more on using addons here"
```
Comment (nitindermohan) [inline content/docs/concepts/oakestra-extensions/addons.md]:
```suggestion
  description="Read more on how to use addons within Oakestra"
```

File: content/docs/concepts/oakestra-extensions/custom_resources.md [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+{{< callout context="note" title="Benefits" icon="outline/rocket">}}
+- **API Flexibility**: Users can introduce new data types and workflows as needed without modifying the core system.
+- **Separation of Concerns**: By decoupling data storage from active state management, custom resources and controllers provide a modular, maintainable architecture.
+- **Enhanced Automation**: With controllers managing custom resources, Oakestra can automate workflows and maintain system alignment with external configurations, streamlining operations.
+{{< /callout >}}
+
+{{< link-card
+  title="Creating Custom Resources"
+  description="Start using custom resources"
```
Comment (nitindermohan) [inline content/docs/concepts/oakestra-extensions/custom_resources.md]:
```suggestion
  description="Read more on how to use custom resources within Oakestra"
```

File: content/docs/concepts/oakestra-extensions/hooks.md [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+{{< callout context="note" title="Benefits" icon="outline/rocket">}}
+- **Reduced Coupling**: Hooks allow components to interact without direct dependencies, enabling a cleaner architecture. For instance, the Root Network can receive deployment notifications without hard-coded calls from the System Manager.
+- **Dynamic Reactivity**: Hooks enable services to adapt dynamically based on lifecycle changes, maintaining an up-to-date environment.
+- **Improved System Flow**: By offering asynchronous and synchronous modes, hooks provide a flexible mechanism that suits both immediate and non-blocking response needs, enhancing system performance and modularity.
+{{< /callout >}}
+
+{{< link-card
+  title="Setting up Hooks"
+  description="Read up on setting up hooks here"
```
Comment (nitindermohan) [inline content/docs/concepts/oakestra-extensions/hooks.md]:
```suggestion
  description="Read up on how to set up hooks within Oakestra
```

**Review by nitindermohan — CHANGES_REQUESTED** [review]
Looks good but minor changes are still required

**Review by nitindermohan — APPROVED** [review]
Looks good! Approved

**Comment by nitindermohan** [comment]
Thank you @melkodary and @Mjaethers for the great work on this!

**Comment by melkodary** [comment]
Special thanks to @Mjaethers for helping me on this ❤️

**Pull request merged**